### PR TITLE
Log worker close reason in events

### PR DESCRIPTION
**Pull request by fjetter**
This allows for some introspection on scheduler side about why a worker closed.

There are many places where the reason may come from. Unfortunately, the actual close on scheduler side is triggered when the connection is broken and therefore we cannot cleanly forward this information during close. However, we can simply add another entry to the events log.

### CODE REVIEW

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -4473,3 +4479,16 @@ async def test_scatter_creates_ts(c, s, a, b):
...
+    client=True,
+    nthreads=[],
+    config={"distributed.scheduler.default-task-durations": {"slowinc": 1000}},
+)
+async def test_scale_up_large_tasks(c, s):
+    futures = c.map(slowinc, range(10))
+    while not s.tasks:
+        await asyncio.sleep(0.001)
+
+    assert s.adaptive_target() == 10
```
Comment (hendrikmakait) [inline distributed/tests/test_scheduler.py:4494]: This test feels unrelated. Has this been added to this PR on purpose?

**Review by hendrikmakait — COMMENTED** [review]
Generally LGTM, one test feels out of place.

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (fjetter) [reply distributed/tests/test_scheduler.py:4494]: oops, sorry about that

**Review by hendrikmakait — APPROVED** [review]

**Comment by hendrikmakait** [comment]
CI was _very_ red which made little sense given the changes. Re-running it before merging.

**Comment by fjetter** [comment]
Well, it does make sense and I was affraid of this already

```python-traceback
distributed/worker.py:1524: in close
    self.log_event(self.address, {"action": "closing-worker", "reason": reason})
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Worker 'not-running', name: None, status: init, stored: 0, running: 0/2, ready: 0, comm: 0, waiting: 0>

    @property
    def address(self) -> str:
        """
        The address this Server can be contacted on.
        If the server is not up, yet, this raises a ValueError.
        """
        if not self._address:
            if self.listener is None:
>               raise ValueError("cannot get address of non-running Server")
E               ValueError: cannot get address of non-running Server
```

This is a problem that pops up if a worker closes before the server is up properly or the server startup failed for whatever reason. we'll also close the worker in this event to make sure we're cleaning up properly but we cannot send anything int his case, of course

**Comment by hendrikmakait** [comment]
> Well, it does make sense and I was affraid of this already

Yup, just realized this as well after diving deeper into the errors after CI didn't recover.

**Comment by hendrikmakait** [comment]
The failures of `test_clear_events_worker_removal` _feel_ new, but at this point, it's hard to tell what's already broken on CI.
